llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp

```diff
@@ -198,48 +202,118 @@ LogicalResult LinalgSCFTilingPattern::returningMatchAndRewrite(
-      filter(std::move(f)), options(std::move(options)) {}
-
-LinalgSCFTileAndFusePattern::LinalgSCFTileAndFusePattern(
-    StringRef opName, MLIRContext *context, scf::SCFTileAndFuseOptions options,
-    LinalgExt::LinalgTransformationFilter f, PatternBenefit benefit)
-    : OpInterfaceRewritePattern<TilingInterface>(context, benefit),
-      filter(f.addOpNameFilter(opName)), options(std::move(options)) {}
+/// Starting from `op` walk all operands backwards to find all
+/// potentially fusable operations, i.e. operations that implement
+/// the `TilingInterface`.
+llvm::SmallDenseSet<Operation *> static collectTiledAndFusedOps(Operation *op) {
+  SmallVector<Operation *> worklist;
+  llvm::SmallDenseSet<Operation *> producers;
+  worklist.push_back(op);
+  producers.insert(op);
+  while (!worklist.empty()) {
+    Operation *current = worklist.pop_back_val();
+    for (OpOperand &operand : current->getOpOperands()) {
+      Operation *producer = operand.get().getDefiningOp();
+      if (!producer || !isa<TilingInterface>(producer) ||
+          producers.count(producer))
+        continue;
+      worklist.push_back(producer);
+      producers.insert(producer);
+    }
+  }
+  return producers;
+}
 
 LogicalResult
-LinalgSCFTileAndFusePattern::matchAndRewrite(TilingInterface op,
-                                             PatternRewriter &rewriter) const {
-  if (failed(filter.checkAndNotify(rewriter, op)))
+SCFTileAndFusePattern::matchAndRewrite(TilingInterface rootOp,
+                                       PatternRewriter &rewriter) const {
+  if (failed(filter.checkAndNotify(rewriter, rootOp)))
     return failure();
 
-  FailureOr<scf::SCFTileAndFuseResult> tiledResults =
-      tileConsumerAndFuseProducerGreedilyUsingSCFForOp(rewriter, op, options);
-  if (failed(tiledResults))
-    return rewriter.notifyMatchFailure(
-        op,
-        "tileConsumerAndFuseProducerGreedilyUsingSCFForOp failed unexpectedly");
-
-  // Replace all uses of the tiled loop operation.
-  SmallVector<Value> replacements(op->getNumResults());
-  for (auto result : llvm::enumerate(op->getResults())) {
-    auto it = tiledResults->replacements.find(result.value());
-    if (it == tiledResults->replacements.end()) {
-      replacements[result.index()] = result.value();
-    } else {
-      replacements[result.index()] = it->getSecond();
-    }
+  // Collect list of operations that can be tiled and fused.
+  llvm::SmallDenseSet<Operation *> origTiledAndFusedOps =
+      collectTiledAndFusedOps(rootOp);
+  auto isIgnoredUser = [&](Operation *user, scf::ForOp outerMostTiledLoop) {
+    return origTiledAndFusedOps.count(user) || isa<tensor::DimOp>(user) ||
+           outerMostTiledLoop->isAncestor(user);
+  };
+
+  // The rest of this method is similar to
+  // scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp, except that also
+  // yields replacements for values of the fused producer.
+
+  // 1. Tile the consumer.
```

Comment:
ditto: are we able to refactor logics out from TestTileConsumerFuseAndYieldProducerUsingSCFForOp and reuse it here?